content/docs/futures/getting_asynchronous.md

```diff
@@ -0,0 +1,174 @@
+---
+title: "Getting asynchronous"
+weight : 2030
+menu:
+  docs:
+    parent: futures
+---
+
+Futures are all about managing asynchronicity. Implementing a future that
+completes asynchonously requires correctly handling receiving `Async::NotReady`
+from the inner future.
+
+Let's start by implementing a future that establishes a TCP socket with a remote
+peer and extracts the peer socket address, writing it to STDOUT.
+
+```rust
+# #![deny(deprecated)]
+extern crate tokio;
+#[macro_use]
+extern crate futures;
+
+use tokio::net::{TcpStream, tcp::ConnectFuture};
+use futures::{Future, Async, Poll};
+use std::io;
+use std::net::SocketAddr;
+
+struct GetPeerAddr {
+    connect: ConnectFuture,
+}
+
+impl Future for GetPeerAddr {
+    type Item = ();
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        match self.connect.poll() {
+            Ok(Async::Ready(socket)) => {
+                println!("peer address = {}", socket.peer_addr().unwrap());
+                Ok(Async::Ready(()))
+            }
+            Ok(Async::NotReady) => Ok(Async::NotReady),
+            Err(e) => {
+                println!("failed to connect");
+                Ok(Async::Ready(()))
+            }
+        }
+    }
+}
+
+fn main() {
+    let addr = "192.168.0.1:1234".parse().unwrap();
+    let connect_future = TcpStream::connect(&addr);
+    let get_peer_addr = GetPeerAddr {
+        connect: connect_future,
+    };
+
+# let get_peer_addr = futures::future::ok::<(), ()>(());
+
+    tokio::run(get_peer_addr);
+}
+```
+
+The implementation of `GetPeerAddr` is very similar to the `Display` future from
+the previous page. The primary difference is, in this case,
+`self.connect.poll()` will (probably) return `Async::NotReady` a number of times
+before returning the connected socket. When this happens, our future returns
+`NotReady`.
+
+`GetPeerAddr` constains [`ConnectFuture`], a future that completes once a TCP
+stream has been established. This future is returned by [`TcpStream::connect`].
+
+When `GetPeerAddr` is passed to `tokio::run`, Tokio will repeatedly call `poll`
+until `Ready` is returned. The exact mechanism by which this happens is
+described in later chapters.
+
+When implementing `Future`, `Async::NotReady` **must not** be returned **unless**
```

Comment:
Having this note here is good. Could you also link to another section that goes into greater depth about what happens if you get this wrong?